client/service/service.go

```diff
@@ -0,0 +1,152 @@
+package boxservice
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"reflect"
+	runtimeDebug "runtime/debug"
+	"sync"
+	"time"
+	"unsafe"
+
+	box "github.com/sagernet/sing-box"
+	"github.com/sagernet/sing-box/adapter"
+	"github.com/sagernet/sing-box/common/urltest"
+	"github.com/sagernet/sing-box/experimental"
+	"github.com/sagernet/sing-box/experimental/clashapi"
+	"github.com/sagernet/sing-box/experimental/libbox"
+	"github.com/sagernet/sing-box/experimental/libbox/platform"
+	"github.com/sagernet/sing-box/option"
+	"github.com/sagernet/sing/service"
+	"github.com/sagernet/sing/service/pause"
+
+	"github.com/getlantern/radiance/client/boxoptions"
+	"github.com/getlantern/radiance/protocol"
+)
+
+// BoxService is a wrapper around libbox.BoxService
+type BoxService struct {
+	*libbox.BoxService
+	ctx      context.Context
+	cancel   context.CancelFunc
+	instance *box.Box
+
+	pauseManager pause.Manager
+	pauseAccess  sync.Mutex
+	pauseTimer   *time.Timer
+}
+
+// New creates a new BoxService that wraps a [libbox.BoxService]. platformInterface is used
+// to interact with the underlying platform
+func New(logOutput string, platIfce platform.Interface) (*BoxService, error) {
+	inboundRegistry, outboundRegistry, endpointRegistry := protocol.GetRegistries()
+	ctx := box.Context(
+		context.Background(),
+		inboundRegistry,
+		outboundRegistry,
+		endpointRegistry,
+	)
+
+	options := boxoptions.Options(logOutput)
+	return newlibbox(ctx, options, platIfce)
+}
+
+// newlibbox creates a new libbox.BoxService instance using the provided context, options, and
+// platformInterface. This should only be used until sing-box is updated to allow wrapping a box.Box
+// instance (if that ever happens).
+func newlibbox(ctx context.Context, options option.Options, platIfce platform.Interface) (*BoxService, error) {
+	////////////////////////////////////////////////////////////////////////
+	// Do not modify the following code unless you know what you're doing //
+	////////////////////////////////////////////////////////////////////////
+
+	experimental.RegisterClashServerConstructor(clashapi.NewServer)
+
+	lbService := new(libbox.BoxService)
+	lbctxptr := getFieldPtr[context.Context](lbService, "ctx")
+
+	// TODO: Do we want to use the filemanager service?
+	//
+	//	ctx = filemanager.WithDefault(ctx, sWorkingPath, sTempPath, sUserID, sGroupID)
+	// ctx = service.ContextWith(ctx, filemgr)
+
+	ctx, cancel := context.WithCancel(ctx)
+	urlTestHistoryStorage := urltest.NewHistoryStorage()
+	ctx = service.ContextWithPtr(ctx, urlTestHistoryStorage)
+
+	if platIfce != nil {
+		service.MustRegister[platform.Interface](ctx, platIfce)
```

Comment:
I noticed you are registering the raw platform interface from Kotlin/Swift directly here. [libbox.NewService](https://github.com/SagerNet/sing-box/blob/dev-next/experimental/libbox/service.go#L57) wraps the platform.Interface in a platformWrapper, which adds extra logic for handling things like [opening the tun device](https://github.com/SagerNet/sing-box/blob/dev-next/experimental/libbox/service.go#L148) and platform-specific behavior. I think not using something like this might break tunneling on Android?